File: 014 - Longest Collatz sequence/C++/pe_014.cc
/*
 * file:    pe_014.cc
 * title:   Longest Collatz sequence
 * author:  Alexander P. Gosselin
 * e-mail:  alexandergosselin@gmail.com
 * date:    October 3, 2014 (modified)
 *          Original date unknown
 */

#include <array>
#include <vector>
#include <iostream>
using namespace std;

const unsigned int N = 999999;

int main() {
  array<int, N> seq_l;
  seq_l.fill(0);
  int max_l = 0;
  int n_max = 0;
  vector<int> add_l;
  vector<int>::iterator it;
  unsigned int i;
  unsigned long long int n;
  for (i = 1; i < N; i++) {
    if (seq_l[i] == 0) {
      n = i + 1;
      add_l.push_back(i);
      while (n > 1) {
        if (n % 2 == 0) {
          n = n/2;
        } else {
          n = 3*n + 1;
        }
        for (it = add_l.begin(); it < add_l.end(); ++it) {
          seq_l[*it]++;
        }
        if (n <= N) {
          if (seq_l[n-1] == 0) {
             add_l.push_back(n-1);
          } else {
            for (it = add_l.begin(); it < add_l.end(); ++it) {
              seq_l[*it] += seq_l[n-1];
            }
            break;
          }
        }
      }
      add_l.erase(add_l.begin(), add_l.end());
      if (seq_l[i] > max_l) {
        max_l = seq_l[i];
        n_max = i + 1;
      }
    } else continue;
  }
  cout << n_max << '\n';
  return 0;
}

File: 014 - Longest Collatz sequence/Haskell/pe_014.hs
{-
file:   pe_014.hs
title:  Longest Collatz sequence
author: Alexander P. Gosselin
e-mail: alexandergosselin@gmail.com
date:   May 23, 2015

link:   https://projecteuler.net/problem=14

note:   This solution is based on the fourth solution to Problem 14
    posted by user Muhkuh at wiki.haskell.org.
    <https://wiki.haskell.org/Euler_problems/11_to_20#Problem_14>
    <https://wiki.haskell.org/User:Muhkuh>
-}

import Data.Array
import Data.List  (maximumBy)
import Data.Ord   (comparing)

collatzLengths :: Int -> Array Int Int
collatzLengths max = collatzLengths'
  where
    collatzLengths' = listArray (0, max) $ 0:map collatzLength [1..]
    collatzLength n
        | n == 1    = 0
        | n' <= max = 1 + collatzLengths' ! n'
        | otherwise = 1 + collatzLength n'
      where
        n' | even n = quot n 2
           | odd n  = 3*n + 1

maxIndex :: Array Int Int -> Int
maxIndex = fst . maximumBy (comparing snd) . assocs

pe_014 = maxIndex $ collatzLengths 1000000

main :: IO ()
main = do
    print pe_014
